### PR TITLE
Either SMT so we can gracefully handle SMT generation issues such as CopySlice

### PR DESCRIPTION
**Pull request by msooseth**
## Description
This allows us to give results even for cases where *some* of the SMT cannot be generated due to symbolic copyslice. Then, we collect the errors and we continue running. This allows us to be more user-friendly and deal with more cases.

Please simply update (i.e. push into) this PR any changes you'd like to see about it!

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [x] updated the changelog

### CODE REVIEW

File: src/EVM/Solvers.hs [hunk 1]
```diff
@@ -139,15 +137,15 @@ withSolvers solver count threads timeout cont = do
...
               -- otherwise call (check-sat), parse the result, and send it down the result channel
               Right () -> do
                 sat <- sendLine inst "(check-sat)"
                 res <- do
                     case sat of
                       "unsat" -> pure Unsat
-                      "timeout" -> pure Unknown
-                      "unknown" -> pure Unknown
+                      "timeout" -> pure $ Timeout "Result timeout by SMT solver"
+                      "unknown" -> pure $ Timeout "Result unknown by SMT solver"
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Solvers.hs:148]: Yeah, it's a bit misleading. I think I wanted to rename it to Unknown.... I'll work on this!

File: src/EVM/Solvers.hs [hunk 2]
```diff
@@ -65,30 +65,28 @@ data Task = Task
 data CheckSatResult
   = Sat SMTCex
   | Unsat
-  | Unknown
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Solvers.hs:68]:
```
data CheckSatResult
  = Sat SMTCex
  | Unsat
  | Unknown String
  | Error String
  deriving (Show, Eq)
```

And that should be OK

File: src/EVM/Solvers.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Solvers.hs]: Renamed all to Unknown to remove this issue

File: src/EVM.hs [hunk 3]
```diff
@@ -663,8 +663,17 @@ exec1 = do
...
+                _ -> do
+                  -- symbolic, ignore gas
+                  next
+                  m <- gets (.state.memory)
+                  case m of
+                    ConcreteMemory mem -> do
+                      buf <- freezeMemory mem
+                      copyBytesToMemory buf sz srcOff dstOff
+                    SymbolicMemory mem -> do
+                      assign (#state % #memory) (SymbolicMemory $ copySlice srcOff dstOff sz mem mem)
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM.hs]: Restructured to reuse, thanks for the idea! :)

**Review by blishko — APPROVED** [review]
LGTM!

File: cli/cli.hs [hunk 4]
```diff
@@ -260,8 +260,8 @@ equivalence cmd = do
     case any isCex res of
       False -> liftIO $ do
         putStrLn "No discrepancies found"
-        when (any isTimeout res) $ do
-          putStrLn "But timeout(s) occurred"
+        when (any isTimeout res || any isError res) $ do
+          putStrLn $ "But " <> (show $ length $ filter isTimeout res) <> " timeout(s) and/or " <> (show $ length $ filter isError res) <> " error(s)/incomplete execution(s) occurred"
```
[the remark this answers is not in the document]
Comment (msooseth) [reply cli/cli.hs]: I categorised these now.

File: src/EVM/UnitTest.hs [hunk 5]
```diff
@@ -240,6 +244,7 @@ symRun opts@UnitTestOptions{..} vm (Sig testName types) = do
       let y = symFailure opts testName (fst cd) types x
       liftIO $ putStr $ "   \x1b[31m[FAIL]\x1b[0m " <> Text.unpack testName <> "\n" <> Text.unpack y
       pure False
+    where
```
Comment (blishko) [inline src/EVM/UnitTest.hs:247]: Looks like trailing `where`?
I am surprised this compiles :D

Comment (msooseth) [reply src/EVM/UnitTest.hs:247]: Wow, yes, thanks, nice catch!

Comment (msooseth) [reply src/EVM/UnitTest.hs]: Fixed.

**Pull request merged**